modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1249 @@
+  large enough to fill the cache line. 
+*/
+module UnrolledLinkedList {
+  import ChapelLocks;
+  private use HaltWrappers;
+  private use Sort;
+  private use IO;
+  private use List;
+
+  pragma "no doc"
+  private param _sanityChecks = true;
+
+  //
+  // Some asserts are useful while developing, but can be turned off when the
+  // implementation is correct.
+  //
+  private inline proc _sanity(expr: bool) {
+    if _sanityChecks then
+      assert(expr);
+  }
+
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let unrolledLinkedList methods have a const ref receiver even
+  // when `parSafe` is `true` and the unrolledLinkedList lock is used.
+  //
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  /* Check that element type is supported by unrolledLinkedList */
+  pragma "no doc"
+  proc _checkType(type eltType) {
+    if !isDefaultInitializable(eltType) {
+      compilerError("unrolledLinkedList element type must be default-initializable");
+    }
+  }
+
+  pragma "no doc"
+  proc _checkNodeCapacity(cap: int) {
+    if cap < 2 {
+      halt("nodeCapacity can't be less than 2");
+    }
+    if cap % 2 == 1 {
+      halt("nodeCapacity can't be odd");
+    }
+  }
+
+  pragma "no doc"
+  class _linkedNode {
+    type eltType;
+    const capacity: int = 32;
+    var size: int = 0;
+    var data: [0..#capacity] eltType;
+    var next: unmanaged _linkedNode(eltType)? = nil;
+
+    proc append(in x: eltType) lifetime this < x {
+      _sanity(size != capacity);
+      data[size] = x;
+      size += 1;
+    }
+
+    proc append(ref x: eltType) where isOwnedClass(x)
+    lifetime this < x {
+      _sanity(size != capacity);
+      data[size] = x;
+      size += 1;
+    }
+  };
+
+  record unrolledLinkedList {
+
+    /* The type of the elements contained in this unrolledLinkedList. */
+    type eltType;
+
+    /* If `true`, this unrolledLinkedList will perform parallel safe operations. */
+    param parSafe = false;
+
+    /* The capacity of one linked node in the unrolledLinkedList */
+    var nodeCapacity: int = 32;
+
+    pragma "no doc"
+    var _size = 0;
+
+    pragma "no doc"
+    var _lock$ = if parSafe then new _LockWrapper() else none;
+
+
+    pragma "no doc"
+    var _head: unmanaged _linkedNode(eltType)? = nil;
+
+    pragma "no doc"
+    var _tail: unmanaged _linkedNode(eltType)? = nil;
+
+    /*
+      Delete all nodes in the list
+    */
+    pragma "no doc"
+    proc _destroy() {
+      var cur = _head;
+      while cur != nil {
+        var next = cur!.next;
+        delete cur;
+        cur = next;
+      }
+    }
+
+    pragma "no doc"
+    proc deinit() {
+      _destroy();
+    }
+
+    /*
+      Initializes an empty unrolledLinkedList.
+
+      :arg eltType: The type of the elements of this unrolledLinkedList.
```

Comment:
I don't like "unrolledLinkedList" in plain English sentences. I wouldn't be opposed if we just said "... of this list". This is a list, after all. Unrolled Linked List is the underlying implementation of it and we shouldn't bend over backwards to repeat that.

Reply:
I agree with you and it does sound awkward in an English sentence. But I'm also afraid that replacing all `unrolledLinkedList` to `list` could be ambiguous because we have a real `List`.